vochain/types.go

```diff
@@ -59,13 +64,73 @@ type QueryData struct {
 	ProcessType string `json:"type,omitempty"`
 }
 
+// ________________________ TRANSACTION COSTS __________________________
+// TransactionCosts describes how much each operation should cost
+type TransactionCosts struct {
+	SetProcess            uint64 `json:"Tx_SetProcess" state:"c_setProcess"`
+	RegisterKey           uint64 `json:"Tx_RegisterKey" state:"c_registerKey"`
+	NewProcess            uint64 `json:"Tx_NewProcess" state:"c_newProcess"`
+	SendTokens            uint64 `json:"Tx_SendTokens" state:"c_sendTokens"`
+	SetAccountInfo        uint64 `json:"Tx_SetAccountInfo" state:"c_setAccountInfo"`
+	AddDelegateForAccount uint64 `json:"Tx_AddDelegateForAccount" state:"c_addDelegateForAccount"`
+	DelDelegateForAccount uint64 `json:"Tx_DelDelegateForAccount" state:"c_delDelegateForAccount"`
+	CollectFaucet         uint64 `json:"Tx_CollectFaucet" state:"c_collectFaucet"`
+}
+
+// StructAsBytes returns the contents of TransactionCosts as a map. Its purpose
+// is to keep knowledge of TransactionCosts' fields within itself, so the
+// function using it only needs to iterate over the key-values.
+func (t *TransactionCosts) StructAsBytes() (b map[string][]byte, err error) {
+	b = make(map[string][]byte)
+
+	tType := reflect.TypeOf(*t)
+	tValue := reflect.ValueOf(*t)
+	for i := 0; i < tType.NumField(); i++ {
+		key := tType.Field(i).Tag.Get("state")
+		value := tValue.Field(i).Uint()
+		valueBytes, err := util.Uint64ToBytes(value)
+		if err != nil {
+			return nil, err
+		}
+		b[key] = valueBytes
+	}
+	return
+}
+
+// TransactionCostsFieldFromStateKey transforms "c_setProcess" to "SetProcess" for all of
+// TransactionCosts' fields
+func TransactionCostsFieldFromStateKey(key string) (string, error) {
+	if len(key) < 3 {
+		return "", fmt.Errorf("state key must have a length greater than 3, because it should include the costPrefix: got %s", key)
+	}
+	if key[0:2] != costPrefix {
+		return "", fmt.Errorf("state keys must start with '%s', got %s", costPrefix, key)
+	}
```

Comment:
that's convenient

Reply:
> you can even reuse the result of TrimLeft, to avoid checking the string twice:

No, this will mean you have to carefully design the order in which errors are returned otherwise edge cases could make the errors not make sense.
Consider `TransactionCostsFieldFromStateKey("c_")` which will result in the error `"" needs to be alphabetical only`, because `name=""` after `strings.TrimPrefix()`.

Better to just check a string several times, what's the hurry? My dad calls this "trying to be smart".

Reply:
is this solved ?